aarch64: use virtual counter

b-man opened the pull request:
The physical counter is mostly reserved for hypervisors.

Any non-secure access to CNTPCT_EL0 may be trapped to EL2 if EL2 is enabled by the current security state, unless CNTHCTL_EL2.EL0PCTEN = 1.

See p. 7859 and p. 7933 of the ARM ARM for more details.

richlowe commented:
I think this needs to be more complicated than it is, the code as you have it hangs on boot in qemu with the boot CPU going idle but never waking up.  It looks like the clock (or any cyclic) is not actually running anymore.  `cbe_hres_tick` never fires.

The register is reading with believably increasing values, but as best as I can tell we're not getting clocks firing, it doesn't seem in fact like we're getting the timer interrupts at all.

Should we stick with the physical counter when we have access to EL2?
Is the virtual counter reset state such that it needs to be more completely configured?

b-man commented:
> I think this needs to be more complicated than it is, the code as you have it hangs on boot in qemu with the boot CPU going idle but never waking up. It looks like the clock (or any cyclic) is not actually running anymore. `cbe_hres_tick` never fires.
> 
I see what you mean - I was running this in QEMU and didn't catch that CPU 0 was hanging (despite the other CPUs spinning up). Sorry about that.
> The register is reading with believably increasing values, but as best as I can tell we're not getting clocks firing, it doesn't seem in fact like we're getting the timer interrupts at all.
> 
Indeed - I'm still getting familiar with the generic timer and didn't catch that the interrupt is going to be different for the virtual timer (looking back this should have been obvious and I shouldn't have just skimmed the timer doc).
> Should we stick with the physical counter when we have access to EL2? Is the virtual counter reset state such that it needs to be more completely configured?

I think sticking to the physical timer if running at EL2 makes sense (and is something that FreeBSD seems to do?) - since we would be running in the same context as the hypervisor at this point (if/when we gain bhyve support).
As for the virtual counter reset, it seems that setting the comparator value, clearing the imask bit, and setting the enable bit should be enough to get it going, but I'll take a closer look.

Btw, I have an 8GB RPi 4 arriving in the mail today, so I'll be able to run more comprehensive tests going forward.

richlowe commented on usr/src/psm/stand/boot/aarch64/common/arch_timer.c:
```diff
@@ -0,0 +1,129 @@
+/*
+ * CDDL HEADER START
```
You probably want the header from `usr/src/prototypes/prototype.c` not the old OpenSolaris one.

b-man replied:
Done.

richlowe commented on usr/src/psm/stand/boot/aarch64/common/arch_timer.c:
```diff
@@ -0,0 +1,129 @@
+ *
+ * CDDL HEADER END
+ */
+
```
Are you sure you don't want to add your copyright?

b-man replied:
I meant to add that yesterday but needed to get to get some sleep before work. :-) Done.

richlowe commented on usr/src/psm/stand/boot/aarch64/common/arch_timer.c:
```diff
@@ -0,0 +1,129 @@
+};
+
+static struct arch_timer_ops timer_ops[] = {
+	{ /* phys timer */
```
can we use C99 array init so we have `[TMR_PHYS] = ...` and `[TMR_VIRT] = ...` and not need the comments?

b-man replied:
Done.

richlowe reviewed (commented):
Just some tiny nits.  This boots in regular qemu now, and seems ok there.  I've asked @citrus-it to see if it solves the problem he had with apple hvf, which seems fairly likely.

citrus-it commented:
I test booted this on a Macbook M2 under qemu, but with these options which always failed before:

```
        -machine virt,accel=hvf,highmem=off
        -cpu host
```

It booted up and got me to the recovery shell, and I was able to log in.

b-man commented:
There was also a logic error and possible overflow in arch_timer_udelay() that I've fixed.

b-man commented:
~~Hmm, looks like this needs a bit more work as the rpi4 MMC driver is failing to initialize with this change.~~
Never mind. I somehow corrupted the partition table on my SD card which was causing the failure. Formatting and resetting the partition table fixed it.

b-man commented:
Also restored interrupt masking/unmasking in cbe_init_pre(), cbe_enable(), and cbe_disable(). Without these the IMASK bit was left in an indeterminate state.

b-man commented:
@richlowe This should be all set for another review.

richlowe commented on usr/src/psm/stand/boot/aarch64/common/arch_timer.c:
```diff
@@ -0,0 +1,186 @@
+arch_timer_freq(void)
+{
+	if (timer_freq)
+		goto end;
```
Can we just early `return (timer_freq)` rather than gotoing?

b-man replied:
Done.

richlowe commented on usr/src/uts/aarch64/sys/arch_timer.h:
```diff
@@ -0,0 +1,51 @@
+extern void arch_timer_select(arch_timer_t type);
+extern void arch_timer_udelay(uint32_t us);
+
+#ifdef __plusplus
```
typo `__cplusplus`.  Did hdrchk notice that?

b-man replied:
That was an error on my part; I didn't run cstyle or hdrchk on any of this code.
I've since ran both cstyle and hdrchk to clear out all stylistic warnings.

richlowe commented on usr/src/uts/armv8/io/cbe.c:
```diff
@@ -341,6 +347,8 @@ get_cbe_vector(void)
 					uint32_t *interrupts = __builtin_alloca(len);
 					prom_getprop(timer, "interrupts", (caddr_t)interrupts);
 					int index = (num > 1)? 1: 0;
+					if (index == 1 && cpu->cpu_m.mcpu_boot_el == 1)
```
Can you leave a comment saying that this is choosing the physical or virtual counter based on boot EL? Just so we don't forget.

b-man replied:
Done.

richlowe commented on usr/src/uts/armv8/os/arch_timer.c:
```diff
@@ -0,0 +1,177 @@
+arch_timer_freq(void)
+{
+	if (timer_freq)
+		goto end;
```
same comment about `goto` v. early return as in psm

b-man replied:
Done.

richlowe commented on usr/src/uts/armv8/os/arch_timer.c:
```diff
@@ -0,0 +1,177 @@
+	cmn_err(CE_PANIC, "arch_timer_freq: missing frequency");
+
+end:
+	return timer_freq;
```
nit: cstyle should have complained about un-parenthesized return here and elsewhere

b-man replied:
Fixed and done.

richlowe reviewed (commented):
This looks good, except for a small handful of nits and preferences.
Can you rebase this into the form you want it merged, when you do that?  I'm not sure if you want it all in one commit, or if there's a logical breakdown you think would be best.

@citrus-it @hadfl do you want to give this another go around on raspberry pi?

b-man committed:
aarch64: implement arch_timer abstraction

This abstraction allows the kernel to select the physical or
virtual counter depending on the boot EL. If the kernel is booted
at EL2, then the physical counter is used. If the kernel is booted at EL1,
then the virtual counter is used. This frees up the physical counter
for use by a hypervisor if booted at EL2, and prevents traps to EL2
if the kernel is booted at EL1 (from within a hypervisor or otherwise).

For now the bootloader is fixed to the physical counter regardless
of EL (this mirrors behavior seen in other bootloaders like u-boot).

b-man commented:
Ok, everything should be shored up.
